dex/wait/queue_test.go

```diff
@@ -0,0 +1,125 @@
+package wait
+
+import (
+	"context"
+	"sync"
+	"testing"
+	"time"
+)
+
+func TestTaper(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	q := NewTaperingTickerQueue(time.Millisecond, time.Millisecond*10)
+	go q.Run(ctx)
+
+	var last time.Time
+	intervals := make([]time.Duration, 0, 10)
+	var expiration sync.WaitGroup
+	expiration.Add(1)
+
+	q.Wait(&Waiter{
+		Expiration: time.Now().Add(time.Millisecond * 30),
+		TryFunc: func() TryDirective {
+			if last.IsZero() {
+				last = time.Now()
+				return TryAgain
+			}
+			intervals = append(intervals, time.Since(last))
+			last = time.Now()
+			return TryAgain
+		},
+		ExpireFunc: func() {
+			expiration.Done()
+		},
+	})
+
+	expiration.Wait()
+
+	if len(intervals) < 3 {
+		t.Fatalf("only %d intervals", len(intervals))
+	}
+
+	// check that the first interval was close to the expected value.
+	var sum time.Duration
+	for _, i := range intervals[:fullSpeedTicks] {
+		sum += i
+	}
+	avg := sum / fullSpeedTicks
+
+	// It'd be nice to check this tighter than *10, but with the race detector,
+	// there are some unexpectedly long times.
+	if avg < time.Millisecond || avg > time.Millisecond*10 {
+		t.Fatalf("first intervals are out of bound: %s", avg)
+	}
+
+	// Make sure it tapered. Can't use the actual last interval, since that
+	// might be truncated. Use the second from the last.
+	lastInterval := intervals[len(intervals)-2]
+	if lastInterval < time.Millisecond*2 {
+		t.Fatalf("last interval wasn't ~5 ms: %s", lastInterval)
+	}
+}
+
+func TestTaperingQueue(t *testing.T) {
+	q := NewTaperingTickerQueue(time.Millisecond, time.Millisecond*5)
+	q.recalcTimer = make(chan struct{}, 5)
```

Comment:
With this and using `q.waiterMtx` and `q.waiters` near the end of this test, I think we're too much in the habit of reaching into internals in tests.  Ideally we should be able to test just the API in a blackbox using a `package wait_test` test setup.